Starwatch.Core/API/Rest/Route/AnnouncementRoute.cs

```diff
@@ -0,0 +1,387 @@
+using Starwatch.API.Gateway;
+using System.Linq;
+using System.Threading.Tasks;
+using Starwatch.Monitoring;
+using Starwatch.Util;
+using System.Diagnostics;
+
+namespace Starwatch.API.Rest.Route
+{
+    [Route("/announcement", AuthLevel.Admin)]
+    class AnnouncementRoute : RestRoute, IGatewayRoute
+    {
+        const string RESOURCE_LOCKED = "Resource is currently locked. Please wait and try again.";
+        const string NO_MONITOR = "AnnouncementMonitor was not found.";
+        const string MISSING_RESOURCE = "No such announcement.";
+
+        const string ID_MUST_BE_INTEGER = "id must be an integer.";
+        const string ID_MUST_BE_POSITIVE = "id must be greater than or equal to 0.";
+
+        public AnnouncementRoute(RestHandler handler, Authentication authentication) : base(handler, authentication) { }
+        public override Type PayloadType => typeof(AnnouncementPatch);
+        public string GetRouteName() => "/announcement";
+
+        public void SetGateway(EventConnection gateway)
+        {
+            Handler = gateway.API.RestHandler;
+            Authentication = gateway.Authentication;
+        }
+
+        /// <summary>
+        /// Gets the account exists and its admin state.
+        /// </summary>
+        /// <param name="query"></param>
+        /// <returns></returns>
+        public override RestResponse OnGet(Query query)
+        {
+#if DEBUG
+            if (Debugger.IsAttached)
+            {
+                if (query.ContainsKey("dbg"))
+                {
+                    if (query["dbg"] == "post")
+                    {
+                        return OnPost(query, null);
+                    }
+
+                    if (query["dbg"] == "delete")
+                    {
+                        return OnDelete(query);
+                    }
+
+                    if (query["dbg"] == "patch")
+                    {
+                        return OnPatch(query, null);
+                    }
+                }
+            }
+#endif
+            // Grab the configurator from the monitor 
+            AnnouncementMonitor mon = null;
+            bool hasMon = Starbound.TryGetMonitor(out mon);
+
+            if (!hasMon)
+            {
+                return new RestResponse(RestStatus.InternalError, "AnnouncementMonitor was not found.");
+            }
+
+            AnnouncementMonitor.Announcement[] announcements = mon.Announcements;
+
+            if (query.ContainsKey("id"))
+            {
+                // Grabs a specific announcement.
+                int id = 0;
+                bool ok = int.TryParse(query["id"], out id);
+                if (!ok)
+                {
+                    // probably unnecessary
+                    id = 0;
+                }
+
+                if (id < 0)
+                {
+                    return new RestResponse(RestStatus.BadRequest, ID_MUST_BE_POSITIVE);
+                }
+
+                if (announcements.Length > id)
```

Comment:
should be `>=`.
arrays are `0 inclusive` to `Length exclusive`

Reply:
If there are 2 announcements, then the index stops at 1.
Since id is expected to be the index (starting at 0), if announcements.Length >= id, then id could be 2 which would cause IndexOutOfRange.

Should I change the implementation to interpret id to start at 1 / announcements[id - 1]?